plugins/hypervisors/kvm/src/main/java/org/apache/cloudstack/kvm/ha/KvmHaHelper.java

```diff
@@ -0,0 +1,190 @@
+        int problematicHostsRatioAccepted = (int) (hostsInCluster.size() * acceptedProblematicHostsRatio);
+
+        if (problematicHosts > problematicHostsRatioAccepted) {
+            ClusterVO cluster = clusterDao.findById(host.getClusterId());
+            LOGGER.warn(String.format("%s is problematic but HA will not fence/recover due to its cluster [id: %d, name: %s] containing %d problematic hosts (Down, Disconnected, "
+                            + "Alert or Error states). Maximum problematic hosts accepted for this cluster is %d.",
+                    host, cluster.getId(), cluster.getName(), problematicHosts, problematicHostsRatioAccepted));
+            return true;
+        }
+        return false;
+    }
+
+    /**
+     * Returns true if the given Host KVM-HA-Helper is reachable by another host in the same cluster.
+     */
+    protected boolean isHostAgentReachableByNeighbour(Host host) {
+        List<HostVO> neighbors = resourceManager.listHostsInClusterByStatus(host.getClusterId(), Status.Up);
+        for (HostVO neighbor : neighbors) {
+            boolean isVmActivtyOnNeighborHost = isKvmHaAgentHealthy(neighbor);
+            if (isVmActivtyOnNeighborHost) {
+                boolean isReachable = kvmHaAgentClient.isHostReachableByNeighbour(neighbor, host);
+                if (isReachable) {
+                    String.format("%s is reachable by neighbour %s. If CloudStack is failing to reach the respective host then it is probably a network issue between the host "
+                            + "and CloudStack management server.", host, neighbor);
+                    return true;
+                }
+            }
+        }
+        return false;
+    }
+
+    /**
+     * Returns true if the host is healthy. The health-check is performed via HTTP GET request to a service that retrieves Running KVM instances via Libvirt. <br>
+     * The health-check is executed on the KVM node and verifies the amount of VMs running and if the Libvirt service is running.
+     */
+    public boolean isKvmHealthyCheckViaLibvirt(Host host) {
+        boolean isKvmHaAgentHealthy = isKvmHaAgentHealthy(host);
+        if (!isKvmHaAgentHealthy && (isClusteProblematic(host) || isHostAgentReachableByNeighbour(host))) {
+            return true;
+        }
+        return isKvmHaAgentHealthy;
+    }
+
+    /**
+     * Checks if the KVM HA webservice is enabled. One can enable or disable it via global settings 'kvm.ha.webservice.enabled'.
+     */
+    public boolean isKvmHaWebserviceEnabled(Host host) {
+        boolean isKvmHaWebserviceEnabled = KVMHAConfig.IsKvmHaWebserviceEnabled.value();
+        if (!isKvmHaWebserviceEnabled) {
+            LOGGER.debug(String.format("Skipping KVM HA web-service verification for %s due to 'kvm.ha.webservice.enabled' not enabled.", host));
+            return false;
+        }
+        return true;
```

Comment:
We could remove both returns and just add a `return isKvmHaWebserviceEnabled;` at the end.

Reply:
@GutoVeronezi these slipped my eyes.
Thanks a lot for bringing it. I will soon update the code.